sys/pool_alloc.c

```diff
@@ -0,0 +1,235 @@
+#include <stdc.h>
+#include <vm.h>
+#include <physmem.h>
+#include <queue.h>
+#include <bitstring.h>
+#include <common.h>
+#include <pool_alloc.h>
+
+#define PI_MAGIC_WORD 0xcafebabe
+
+#define PP_FREED_WORD 0xdeadbeef
+
+#define GET_PI_AT_IDX(slab, i, size)                                           \
+  (pool_item_t *)((unsigned long)slab + (slab->ph_start) +                     \
+                  (i) * (size + sizeof(pool_item_t)))
+
+#define ALIGN_TO_BINARY_POW(x, a) ((((x) >> (a)) + 1) << (a))
+
+#define PALLOC_DEBUG 0
+
+#if defined(PALLOC_DEBUG) && PALLOC_DEBUG > 0
+#define PALLOC_DEBUG_PRINT(text, args...)                                      \
+  kprintf("PALLOC_DEBUG: %s:%d:%s(): " text, __FILE__, __LINE__, __func__,     \
+          ##args)
+#else
+#define PALLOC_DEBUG_PRINT(text, args...)
+#endif
+
+typedef struct pool_item {
+  unsigned long pi_guard_number; /* PI_MAGIC_WORD by default, normally isn't */
+  /* changed, so if it has another value, memory */
+  /* corruption has taken place */
+  pool_slab_t *pi_slab; /* pointer to slab containing this item */
+  unsigned long pi_data[0];
+} pool_item_t;
+
+pool_slab_t *create_slab(size_t size, pool_t *pool,
+                         void (*constructor)(void *, size_t)) {
+  PALLOC_DEBUG_PRINT("Entering create_slab\n");
+  vm_page_t *page_for_slab = pm_alloc(1);
+  pool_slab_t *slab = (pool_slab_t *)page_for_slab->vaddr;
+  slab->ph_page = page_for_slab;
+  slab->ph_nused = 0;
+  slab->ph_ntotal =
+    (PAGESIZE - sizeof(pool_slab_t) - 3) /
+    (sizeof(pool_item_t) + size +
+     1); /* sizeof(pool_slab_t)+n*(sizeof(pool_item_t)+size+1)+7+sizeof(unsigned
+          long)
+   <= PAGESIZE, 7 is maximum number of padding bytes
+   for a bitmap */
+
+  slab->ph_nfree = slab->ph_ntotal;
+  slab->ph_start =
+    sizeof(pool_slab_t) + ALIGN_TO_BINARY_POW(slab->ph_ntotal, 2);
+  memset(slab->ph_bitmap, 0, slab->ph_ntotal);
```

Comment:
I think you are clearing too much with this memset, the size of a bitmap with n entries is `bitstr_size(n)`.